processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByMergingQueryRunnerV2.java

```diff
@@ -224,53 +231,83 @@ public GroupByMergingQueryRunnerV2(
-                                    throw new ISE(
-                                        "Null queryRunner! Looks to be some segment unmapping action happening"
-                                    );
-                                  }
+                  final ReferenceCountingResourceHolder<Grouper<GroupByMergingKey>> grouperHolder = new ReferenceCountingResourceHolder<>(
+                      grouper,
+                      new Closeable()
+                      {
+                        @Override
+                        public void close() throws IOException
+                        {
+                          grouper.close();
+                        }
+                      }
+                  );
 
-                                  return exec.submit(
-                                      new AbstractPrioritizedCallable<Void>(priority)
-                                      {
-                                        @Override
-                                        public Void call() throws Exception
-                                        {
-                                          try {
-                                            input.run(queryForRunners, responseContext)
-                                                 .accumulate(grouper, accumulator);
-                                            return null;
-                                          }
-                                          catch (QueryInterruptedException e) {
-                                            throw Throwables.propagate(e);
-                                          }
-                                          catch (Exception e) {
-                                            log.error(e, "Exception with one of the sequences!");
-                                            throw Throwables.propagate(e);
+                  try {
+                    ListenableFuture<List<Boolean>> futures = Futures.allAsList(
+                        Lists.newArrayList(
+                            Iterables.transform(
+                                queryables,
+                                new Function<QueryRunner<Row>, ListenableFuture<Boolean>>()
+                                {
+                                  @Override
+                                  public ListenableFuture<Boolean> apply(final QueryRunner<Row> input)
+                                  {
+                                    if (input == null) {
+                                      throw new ISE(
+                                          "Null queryRunner! Looks to be some segment unmapping action happening"
+                                      );
+                                    }
+
+                                    final ReferenceCountingResourceHolder.Releaser bufferReleaser = mergeBufferHolder.increment();
+                                    final ReferenceCountingResourceHolder.Releaser grouperReleaser = grouperHolder.increment();
```

Comment:
grouperHolder.increment() can throw ISE here, causing the mergeBuffer to remain unclosed ?